### ISSUE TITLE
Add option to pass content as widget instead of plain text

### ISSUE THREAD

**Issue by guenth39** [talamaska/onboarding_overlay#21]
So far I absolutely love the package. Thank you so much for this! Unfortunately I really need the function to also display custom widgets in the overlay. Otherwise you can't display images and animations. Without this function, the package is unfortunately only partly useful for me 😕
For me it could be either a simple widget that you can pass in instead of title and bodyText or a named constructor with other options or a builder, then you could even combine it with title and bodyText.

Best Regards

**Comment by talamaska** [comment]
I'll see what I can do.

**Comment by talamaska** [comment]
@guenth39 There is one thing you should know, the texts are placed in a limited size box. if you try to put something bigger there you will get an overflow error. The not full-screen overlays provide significantly smaller available space. I'm not going to provide a way to change the size of the placeholder, because from it depends the position and a bunch of other stuff that I calculate based on the position of the focused widget and the size of the box.

**Comment by talamaska** [comment]
@guenth39 check this branch if that suits your needs. added a demo in the example folder.
https://github.com/talamaska/onboarding_overlay/tree/feature/provide-builder

**Comment by guenth39** [comment]
First of all, thank you for your quick investigation! Great how quickly you took care of my request.
The current changes on the branch are already an improvement for me in any case! Since I only want to use fullscreen overlays, it's a bit of pity that I can only use the small box. If you want to explain something to an icon in the AppBar then a lot of space on the screen is lost that you could actually use. Could you not make a differentiation here or insert a parameter with which you can allow the overflow explicitly? If this does not work, it would be a pity, but the changes are in any case already very helpful for me. Thanks a lot!

Apart from that, I have two other smaller suggestions:
1. Could you maybe add the size to the build function?
2. Would it make sense to create a separate model `RenderInfo` or something like that? Then you don't have to take all parameters even if you don't need them. Five or with size six parameters are a lot.  And should something be added, is the refactoring effort less if all information are passed into the builder as one parameter or possibly two, if context remains extra.

What do you think about this?

**Comment by guenth39** [comment]
Oh, or I have another idea that could be a compromise. Because actually the goal for me is to add in the description of single widgets also two fullscreen steps without target/focused widget to integrate simply swipe up/down and swipe left/right explanation. For the rest of the steps I can live with the small box.  
That no widget is cut out of the overlay, I have already achieved by passing in a focus that is not assigned to any widget. Couldn't you make a named constructor `OnboardingStep.withoutFocus` or something like that and this can then only be fullscreen, has a FocusNode that is not assigned to a widget or the FocusNode is null and the size of the box is not limited? This could then look something like this:
``` dart
 const OnboardingStep.withoutFocus({
    this.key,
    required this.title,
    this.titleTextColor = const Color(0xFFFFFFFF),
    this.titleTextStyle,
    this.bodyText = '',
    this.bodyTextStyle,
    this.bodyTextColor = const Color(0xFFFFFFFF),
    this.textAlign = TextAlign.start,
    this.shape = const RoundedRectangleBorder(
      borderRadius: BorderRadius.all(Radius.circular(8.0)),
    ),
    this.margin = const EdgeInsets.all(8.0),
    this.overlayColor = const Color(0xC4000000),
    this.overlayShape = const RoundedRectangleBorder(
      borderRadius: BorderRadius.all(Radius.circular(8.0)),
    ),
    this.hasLabelBox = false,
    this.labelBoxPadding = const EdgeInsets.all(8.0),
    this.labelBoxDecoration = const BoxDecoration(),
    this.delay = Duration.zero,
    this.hasArrow = false,
    this.arrowPosition = ArrowPosition.top,
    this.overlayBehavior = HitTestBehavior.opaque,
    this.stepBuilder,
  })  : 
   fullscreen = true,
   focusNode = _defaultUnassignedFocusNode,
   unlimitedLabelBox = true,
   assert(titleTextColor != null || titleTextStyle != null,
            'You should provide at least one of titleTextColor or titleTextStyle'),
   assert(bodyTextColor != null || bodyTextStyle != null,
            'You should provide at least one of bodyTextColor or bodyTextStyle'),
   assert(
            (hasArrow && hasLabelBox) ||
                (!hasArrow && !hasLabelBox) ||
                !hasArrow && hasLabelBox,
            'hasArrow $hasArrow cannot be true if hasLabelBox $hasLabelBox is false');
```

**Comment by talamaska** [comment]
Actually there is a possibility to have an overlay without focus not on a full screen. Just recently fixed an issue around it. Yes placing the arguments in a separate class/model is a next step optimization. Placing any GestureDetectors in the builder won't work as there is a wrapping GestureDetector that take precedence as it listens to PointerDown/TapDown as I need the position of the tap, so that you would be able to click on the widget in focus. So your idea of having carousel like experience inside one step won't work. I have previously closed a similar request for the same reason. Sorry. 
I'll research how can I make the box bigger and it's size calculated dependent on the space available. 
It's is an onboarding not a tutorial or something. The current size was sufficient mostly. I also would suggest taking a look at https://pub.dev/packages/pal and https://pub.dev/packages/pal_widgets, Maybe it suits better for your needs. I don't plan implementing everything that it has. It's way more advanced than what I was planning for this package.

**Comment by talamaska** [comment]
@guenth39 check latest changes in https://github.com/talamaska/onboarding_overlay/tree/feature/provide-builder

**Comment by talamaska** [comment]
Also maybe your carousel idea is implementable now. Give it a try.

**Comment by talamaska** [comment]
> First of all, thank you for your quick investigation! Great how quickly you took care of my request.

:) Np, I had time to tinker a bit. I appreciate your feedback. I hope those additions will make my package even more flexible.

**Comment by talamaska** [comment]
I'll thing in the future if I can expose the controller to be passable as a parameter to the Onboarding

**Comment by guenth39** [comment]
Just saw it and it looks awesome! Thanks a lot 🎉 That's much more than I expected!

Nevertheless, I have found one small bug. When I insert the following step, only a square area is available, whether theoretically on the whole screen space would be? Can you change this?

The rest works really great! I can't be thankful enough for you. Do you have a buy me a coffee button somewhere?

``` dart
OnboardingStep(
  manualControl: true,
  focusNode: FocusNode(),
  title: 'Try this cool Feature!',
  hasLabelBox: false,
  fullscreen: true,
  hasArrow: false,
  overlayColor: Colors.white.withOpacity(.8),
  titleTextColor: Colors.black,
  overlayBehavior: HitTestBehavior.deferToChild,
  shape: const CircleBorder(),
  margin: const EdgeInsets.all(12),
  stepBuilder:
      (BuildContext context, OnboardingStepRenderInfo renderInfo) =>
          Column(
    children: [
      Text(
        renderInfo.title,
        style: renderInfo.titleStyle,
      ),
      Expanded(
        child: Container(
          color: Colors.green[100],
          child: Center(
            child: Text('Here could be an Image'),
          ),
        ),
      ),
    ],
  ),
),
```

**Comment by talamaska** [comment]
@guenth39 fixed

**Comment by guenth39** [comment]
Just checked it and works perfectly 😊👍🏻 Thank you very much! 
I really think the changes on this branch are very good for the package, as it's more flexible now, like you already mentioned earlier. 
Unprecedented in what speed you have implemented the proposals. I am looking forward to the new version on pub.dev. Thank you! 
And as said before, let me know when you have a "buy me a coffee" link 😉

**Comment by jlnrrg** [comment]
It might me worth to consider to get rid of `titleText`, `bodyText`, etc. and only support a child widget instead of `stepBuilder`
This would mimic the normal behavior of flutter Widgets.
(Warning: if this is implemented, the widget provided, needs to be wrapped in a scaffold, so the `OnboardingStepRenderInfo` is not longer needed to be provided to `Text`.

**Comment by jlnrrg** [comment]
Fair point :smile: 
Just for my understanding... how does the stepBuilder fit in this construct (as it works kind of like a child)?

**Comment by talamaska** [comment]
Kind of, but not exactly, it's a non breaking way of adding custom content for me.

**Comment by talamaska** [comment]
fixed in #23
